DB/src/main/java/io/deephaven/db/v2/by/StreamLastByChunkedOperator.java

```diff
@@ -0,0 +1,254 @@
+package io.deephaven.db.v2.by;
+
+import io.deephaven.base.verify.Assert;
+import io.deephaven.db.tables.Table;
+import io.deephaven.db.tables.select.MatchPair;
+import io.deephaven.db.v2.QueryTable;
+import io.deephaven.db.v2.ShiftAwareListener;
+import io.deephaven.db.v2.sort.permute.PermuteKernel;
+import io.deephaven.db.v2.sort.timsort.LongIntTimsortKernel;
+import io.deephaven.db.v2.sources.*;
+import io.deephaven.db.v2.sources.chunk.Attributes.ChunkLengths;
+import io.deephaven.db.v2.sources.chunk.Attributes.ChunkPositions;
+import io.deephaven.db.v2.sources.chunk.Attributes.KeyIndices;
+import io.deephaven.db.v2.sources.chunk.Attributes.Values;
+import io.deephaven.db.v2.sources.chunk.*;
+import io.deephaven.db.v2.utils.ChunkUtils;
+import io.deephaven.db.v2.utils.Index;
+import io.deephaven.db.v2.utils.OrderedKeys;
+import io.deephaven.db.v2.utils.ReadOnlyIndex;
+import io.deephaven.util.SafeCloseableList;
+import org.jetbrains.annotations.NotNull;
+
+import java.lang.ref.SoftReference;
+import java.util.Arrays;
+import java.util.Collections;
+import java.util.LinkedHashMap;
+import java.util.Map;
+
+/**
+ * A lastBy aggregation operator for stream tables.
+ *
+ * @see Table#STREAM_TABLE_ATTRIBUTE
+ */
+public class StreamLastByChunkedOperator implements IterativeChunkedAggregationOperator {
+
+    private static final int COPY_CHUNK_SIZE = ArrayBackedColumnSource.BLOCK_SIZE;
+
+    private final int numResultColumns;
+    private final ColumnSource<?>[] inputColumns;
+    private final WritableSource<?>[] outputColumns;
+    private final PermuteKernel[] permuteKernels;
+
+    private final Map<String, ArrayBackedColumnSource<?>> resultColumns;
+
+    private SoftReference<LongArraySource> cachedRedirections;
+    private LongArraySource redirections;
+
+    StreamLastByChunkedOperator(@NotNull final MatchPair[] resultPairs, @NotNull final Table streamTable) {
+        numResultColumns = resultPairs.length;
+        inputColumns = new ColumnSource[numResultColumns];
+        outputColumns = new WritableSource[numResultColumns];
+        permuteKernels = new PermuteKernel[numResultColumns];
+        final Map<String, ArrayBackedColumnSource<?>> resultColumnsMutable = new LinkedHashMap<>(numResultColumns);
+        for (int ci = 0; ci < numResultColumns; ++ci) {
+            final MatchPair resultPair = resultPairs[ci];
+            final ColumnSource<?> streamSource = streamTable.getColumnSource(resultPair.left());
+            final ArrayBackedColumnSource<?> resultSource = ArrayBackedColumnSource.getMemoryColumnSource(0, streamSource.getType(), streamSource.getComponentType());
+            resultColumnsMutable.put(resultPair.left(), resultSource);
+            inputColumns[ci] = ReinterpretUtilities.maybeConvertToPrimitive(streamSource);
+            outputColumns[ci] = (WritableSource<?>) ReinterpretUtilities.maybeConvertToPrimitive(resultSource);
```

Comment:
Again my backwards assertion.

Reply:
See note in https://github.com/deephaven/deephaven-core/pull/977#discussion_r684845459